File: python_func/python_func.c
#include <uapi/linux/ptrace.h>

// 定义数据结构和性能事件映射
struct data_t {
    char filename[128];
    char funcname[64];
    int lineno;
};
BPF_PERF_OUTPUT(events);

int print_functions(struct pt_regs *ctx)
{
    uint64_t argptr;
    struct data_t data = { };

    // 参数1是文件名
    bpf_usdt_readarg(1, ctx, &argptr);
    bpf_probe_read_user(&data.filename, sizeof(data.filename), (void *)argptr);

    // 参数2是函数名
    bpf_usdt_readarg(2, ctx, &argptr);
    bpf_probe_read_user(&data.funcname, sizeof(data.funcname), (void *)argptr);

    // 参数3是行号
    bpf_usdt_readarg(3, ctx, &data.lineno);

    // 最后提交性能事件
    events.perf_submit(ctx, &data, sizeof(data));
    return 0;
}

File: python_func/python_func.py
import sys
from bcc import BPF, USDT

if len(sys.argv) < 2:
    print("Usage: %s <tracee_pid>" % sys.argv[0])
    sys.exit(1)

u = USDT(pid=int(sys.argv[1]))
u.enable_probe(probe="function__entry", fn_name="print_functions")
b = BPF(src_file="python_func.c", usdt_contexts=[u])


def print_event(data):
    event = b["events"].event(data)
    print("%-9s %-6d %s" % (event.filename, event.lineno, event.funcname))


# 打印头
print("%-9s %-6s %s" % ("FILE_NAME", "LINE_NUMBER", "FUNC_NAME"))

# 绑定性能事件映射和输出函数，并从映射中循环读取数据
b["events"].open_perf_buffer(print_event)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
